README.md
[![Build Status](https://travis-ci.org/sematext/sematext-agent-nginx.svg?branch=master)](https://travis-ci.org/sematext/sematext-agent-nginx)

This is the NGINX monitoring Agent for [Sematext Cloud](https://sematext.com/cloud/)


# Preparation 

1. Get a free [Sematext account](https://apps.sematext.com/ui/registation)  

2. [Create a Monitoring App](https://apps.sematext.com/spm-reports/registerApplication.do) of type "Nginx" and copy the SPM App Token - or execute the commands displayed in the Sematext UI (which are described here as well)

3. [Install Node.js](https://nodejs.org/en/download/package-manager/) on your Nginx server

4. Activate Nginx ```stub_status``` module in the ```server``` section e.g. in ```/etc/nginx/sites-enabled/default```: 
```
location /nginx_status {
  # Turn on nginx stats
  stub_status on;
  # I do not need logs for stats
  access_log   off;
  # Security: Please use you local IP address instead ...
  allow all;
}
```

__Optional preparation for PHP FastCGI Process Manager (FPM):__ To add monitoring for PHP-FPM follow [these instructions](https://sematext.com/docs/integration/php/#integration-with-nginx).

# Setup 
```sh
# Install sematext-agent-nginx 
npm i sematext-agent-nginx -g
# If you use Sematext Cloud EU, set region for API endpoints
# sematext-nginx-setup -r EU
# Install systemd or upstart service file for sematext-agent-nginx 
sematext-nginx-setup -t YOUR_SPM_TOKEN_HERE -n http://localhost/nginx_status
```

# Configuration 

The setup script stores the configuration in ```/etc/sematext/sematext-agent-nginx.config```

In case you want to change settings later edit ```/etc/sematext/sematext-agent-nginx.config```. 
A typical case is to add receiver URL for On-Premises installation of SPM in the config file:

```
# default value for SaaS / Sematext Cloud: https://spm-receiver.sematext.com:443/receiver/v1/_bulk
spmSenderBulkInsertUrl: http://your-spm-server:8084/_bulk
# default value for SaaS / Sematext Cloud: https://event-receiver.sematext.com
eventsReceiverUrl:  http://your-spm-server:8083
```


Restart the Sematext Nginx Agent after config changes, depending on the init system:
- Upstart (Ubuntu):  
```
    sudo service sematext-agent-nginx restart 
```
- Systemd (Linux others):  
```
    sudo systemctl stop sematext-agent-nginx
    sudo systemctl start sematext-agent-nginx
```
- Launchd (Mac OS X): 
```
    sudo launchctl stop com.sematext.sematext-agent-nginx
    sudo launchctl stop com.sematext.sematext-agent-nginx
```

For tests you can just run the agent from command line:
```
sematext-agent-nginx --config /etc/sematext/sematext-agent-nginx.config
```

# Results

NGINX Metrics in Sematext: 
![](https://raw.githubusercontent.com/sematext/sematext-agent-nginx/master/nginx-report-screenshot.png)

# Docker 

Sematext Agent for Nginx includes a docker file and startup script to build a Docker image. 
```
git clone https://github.com/sematext/sematext-agent-nginx.git
cd sematext-agent-nginx
docker build -t sematext/sematext-agent-nginx .
```

The Sematext Nginx Agent supports following parameters on Docker: 

| Environment Variable | Description |
|----------------------|-------------|
| **Required parameters**  |         |
| SPM_TOKEN                | your SPM Token for the Nginx SPM App |
| NGINX_STATUS_URL          | the URL to Nginx server, delivering the stats (see Nginx configuration above). Please note the servername/ip must be reachable from the agent container. You might need to use --link nginx-container-name to create the network link. NGINX_STATUS_URL is not required with DOCKER_AUTO_DISCOVERY.|
|**Docker auto discovery** | Detect new nginx containers for monitoring! |
| DOCKER_AUTO_DISCOVERY | Enable auto discovery of containers e.g. DOCKER_AUTO_DISCOVERY=true |
| SPM_DOCKER_NETWORK | Name of the network to be used for HTTP queries to nginx. If this is set to "host" and docker run parameter ```--net=host``` the connection is made to the exposed ports. If this is not set or any other network name  (e.g. bridge) the connection is done via nginx container IP address and 'internal' port 80 (used inside the nginx container). This feature works only when the Docker socket is mounted with ```-v /var/run/docker.sock:/var/run/docker.sock```| 
| NGINX_STATUS_PATH  | Location of the nginx status page e.g. "/nginx_status" |
| PHP_FPM_STATUS_PATH | Location of the PHP FPM status page e.g. "/status" |
| IMAGE_NAME_PATTERN | Regular expression to match nginx image name. Default  value 'nginx'|
| **General parameters** | |
| HTTPS_PROXY              | Url to HTTPS proxy if the agent runs behind a firewall |
| SPM_RECEIVER_URL         | Optional for SPM On-Premises, default value: https://spm-receiver.sematext.com:443/receiver/v1/_bulk for On-Premises receiver use http://your-spm-server:8084|_bulk.  
| EVENTS_RECEIVER_URL      | Optional for SPM On-Premises, default value: https://event-receiver.sematext.com. For On-Premises use http://your-spm-server:8083|


Example:
```
docker run --name sematext-agent-nginx -e SPM_TOKEN=YOUR_SPM_NGINX_TOKEN_HERE  \ 
-e NGINX_STATUS_URL=http://nginx-server/nginx_status \ 
-d  sematext/sematext-agent-nginx
```

Example with auto discovery of nginx containers via Docker API: 
```
docker run --name sematext-agent-nginx \
-e SPM_TOKEN=YOUR_SPM_NGINX_TOKEN_HERE \
-e NGINX_STATUS_PATH=/nginx_Status \
-e DOCKER_AUTO_DISCOVERY=true \
--net=host -e SPM_DOCKER_NETWORK=host \
-v /var/run/docker.sock:/var/run/docker.sock -d \
sematext/sematext-agent-nginx
```


# Support 

- Twitter: [@sematext](https://www.twitter.com/sematext)
- Blog: [blog.sematext.com](https://blog.sematext.com)
- Homepage: [www.sematext.com](https://www.sematext.com)


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [2.0.4](https://github.com/sematext/sematext-agent-nginx/compare/2.0.3...2.0.4)

> 9 May 2020

- added .npmrc config [`e56b0c6`](https://github.com/sematext/sematext-agent-nginx/commit/e56b0c6af20c623b2abe5dc631b3dc471b472e56)

#### [2.0.3](https://github.com/sematext/sematext-agent-nginx/compare/2.0.2...2.0.3)

> 9 May 2020

- refactored diag package [`81c7641`](https://github.com/sematext/sematext-agent-nginx/commit/81c764142d31d5ff94dcfe0e5becda3f00d09862)
- added release script [`46c2b47`](https://github.com/sematext/sematext-agent-nginx/commit/46c2b47b14fbfc05fa33d63221947e5949728c66)
- Release 2.0.3 [`45b7eef`](https://github.com/sematext/sematext-agent-nginx/commit/45b7eefe8c2c231e553ab9a2abd2e09d84e27d39)

#### [2.0.2](https://github.com/sematext/sematext-agent-nginx/compare/2.0.1...2.0.2)

> 12 December 2019

- fix negative values in charts with comulative metrics [`a159502`](https://github.com/sematext/sematext-agent-nginx/commit/a15950289990e0bbb906ab5a6a7155b4b5303ff1)
- update dependencies [`7620ebe`](https://github.com/sematext/sematext-agent-nginx/commit/7620ebe7a55583b48b94a676eae2d1653c765a1a)
- Release 2.0.2 [`5a122b8`](https://github.com/sematext/sematext-agent-nginx/commit/5a122b8f9277912593989e91b8fd03470409e776)

#### [2.0.1](https://github.com/sematext/sematext-agent-nginx/compare/2.0.0...2.0.1)

> 10 December 2019

- Wording and link tweaks [`1df3687`](https://github.com/sematext/sematext-agent-nginx/commit/1df3687e7db68eaec5e795bfb7b933fde4e05c25)
- update dependencies [`a9b9765`](https://github.com/sematext/sematext-agent-nginx/commit/a9b9765c55c224421872b88f27083f8862dec23c)
- Release 2.0.1 [`4c572d6`](https://github.com/sematext/sematext-agent-nginx/commit/4c572d65cf941b31aa37c09923a73ed6b7320b21)

### [2.0.0](https://github.com/sematext/sematext-agent-nginx/compare/1.30.17...2.0.0)

> 28 December 2018

- update dependencies [`9333a08`](https://github.com/sematext/sematext-agent-nginx/commit/9333a084d0cfd3cf93840fed8d97ce9e3aa26c3c)
- Release 2.0.0 [`cb24da7`](https://github.com/sematext/sematext-agent-nginx/commit/cb24da7a4d539bbf83ba73d2612785f47ff21d2e)

#### [1.30.17](https://github.com/sematext/sematext-agent-nginx/compare/1.30.16...1.30.17)

> 5 September 2018

- Moved to https://sematext.com/docs/integration/php [`350b851`](https://github.com/sematext/sematext-agent-nginx/commit/350b851d91887255a7fe6d29a7faecaac444a516)
- patch request module at runtime [`6c702bd`](https://github.com/sematext/sematext-agent-nginx/commit/6c702bd87bd158e0cc0ebf11e6ad3cdab1896000)
- Release 1.30.17 [`e1e0f1f`](https://github.com/sematext/sematext-agent-nginx/commit/e1e0f1fcef4730ee9e821a56d1232c61411f6d4b)

#### [1.30.16](https://github.com/sematext/sematext-agent-nginx/compare/1.30.15...1.30.16)

> 6 July 2017

- add -t token, -n nginx_url, -p php_url, -r EU|US [`86a42d3`](https://github.com/sematext/sematext-agent-nginx/commit/86a42d30df19eff6cb56d7554d88d3c2a34b37b4)
- Release 1.30.16 [`f785d73`](https://github.com/sematext/sematext-agent-nginx/commit/f785d7350212343e4eeff1f2e06ac8f81fd4921d)
- fix nginx_url -&gt;uppercase [`df92cab`](https://github.com/sematext/sematext-agent-nginx/commit/df92cab38e8d429246d387e7514993a2093e3de8)

#### [1.30.15](https://github.com/sematext/sematext-agent-nginx/compare/1.30.14...1.30.15)

> 26 June 2017

- Release 1.30.15 [`a95e3f7`](https://github.com/sematext/sematext-agent-nginx/commit/a95e3f737636c3f344c7e5b33014ec3e6ca9cfa0)
- update php-fpm instructions [`50ad496`](https://github.com/sematext/sematext-agent-nginx/commit/50ad496472b4718890d822aadf3abc1e60345835)
- Update README.md [`c009cb5`](https://github.com/sematext/sematext-agent-nginx/commit/c009cb51bf362556bf0de1f2bd4621140b0f25eb)

#### [1.30.14](https://github.com/sematext/sematext-agent-nginx/compare/1.30.13...1.30.14)

> 17 March 2017

- Update php-fpm instructions [`9e251a5`](https://github.com/sematext/sematext-agent-nginx/commit/9e251a5d34ec5a10c7af75bb997896025a05988f)
- Release 1.30.14 [`bb17139`](https://github.com/sematext/sematext-agent-nginx/commit/bb1713911f0e7c94dfd8c0eefd37cbc1e2cb81ad)

#### [1.30.13](https://github.com/sematext/sematext-agent-nginx/compare/1.30.12...1.30.13)

> 25 January 2017

- generate default config in YAML format [`ebd8be2`](https://github.com/sematext/sematext-agent-nginx/commit/ebd8be2761213928d80ac352d0f768e90f818744)
- dependency update [`02c4ceb`](https://github.com/sematext/sematext-agent-nginx/commit/02c4ceb294369e8cec8bf698c6d3e55df50eca2f)
- Release 1.30.13 [`b9ecdd1`](https://github.com/sematext/sematext-agent-nginx/commit/b9ecdd102496e8b107e20483856568aaf8b94d08)

#### [1.30.12](https://github.com/sematext/sematext-agent-nginx/compare/1.30.11...1.30.12)

> 23 December 2016

- accept array format in config for phpFpm.url [`ad7f6cf`](https://github.com/sematext/sematext-agent-nginx/commit/ad7f6cfe34929a5c88004606c9c4227be91d07ec)
- Release 1.30.12 [`5f2612b`](https://github.com/sematext/sematext-agent-nginx/commit/5f2612b488ae50e04550b18e204bcb475879317a)

#### [1.30.11](https://github.com/sematext/sematext-agent-nginx/compare/1.30.10...1.30.11)

> 23 December 2016

- Release 1.30.11 [`782908c`](https://github.com/sematext/sematext-agent-nginx/commit/782908cebc93b40c1a3cfbbb6f9622763eb562f8)
- remove debug output [`fb90ab4`](https://github.com/sematext/sematext-agent-nginx/commit/fb90ab47a9579613c1e5f7354045725f5853f7d3)

#### [1.30.10](https://github.com/sematext/sematext-agent-nginx/compare/1.30.9...1.30.10)

> 23 December 2016

- support multiple PHP FPM status urls [`20d9b88`](https://github.com/sematext/sematext-agent-nginx/commit/20d9b88709f398ce931e6ed39195fb2cf1741354)
- Release 1.30.10 [`0045222`](https://github.com/sematext/sematext-agent-nginx/commit/004522272a6f06699634bdf1f0e885aa02ab4c7b)


lib/index.js
#!/bin/sh
':' // ; export MAX_MEM="--max-old-space-size=60"; exec "$(command -v node || command -v nodejs)" "${NODE_OPTIONS:-$MAX_MEM}" "$0" "$@"

/*
 * @copyright Copyright (c) Sematext Group, Inc. - All Rights Reserved
 *
 * @licence Sematext Agent for Nginx is free-to-use, proprietary software.
 * THIS IS PROPRIETARY SOURCE CODE OF Sematext Group, Inc. (Sematext)
 * This source code may not be copied, reverse engineered, or altered for any purpose.
 * This source code is to be used exclusively by users and customers of Sematext.
 * Please see the full license (found in LICENSE in this distribution) for details on its license and the licenses of its dependencies.
 */
// this requires are here to compile with enclose.js
// var packageJson = require('../package.json')
// var packageJson2 = require('spm-agent/package.json')
var autodiscovery = process.env.DOCKER_AUTO_DISCOVERY === 'true' || process.env.DOCKER_AUTO_DISCOVERY === '1' || false
var runningNginxAgents = {}
var statusPath = process.env.NGINX_STATUS_PATH || '/nginx_status'
var phpFpmStatusPath = process.env.PHP_FPM_STATUS_PATH || 'undefined'

function NginxMonitor () {
  var SpmAgent = require('spm-agent')
  var OsAgent = require('spm-agent-os')
  var NginxAgent = require('./nginx-agent')
  console.log('SPM Token: ' + (SpmAgent.Config.get('tokens.spm') || '').slice(0, 16) + ' ...')
  var nginxUrl = null
  var njsAgent = new SpmAgent()
  if (SpmAgent.Config.get('tokens.spm')) {
    try {
      njsAgent.createAgent(new OsAgent())
      if (!autodiscovery) {
        nginxUrl = SpmAgent.Config.get('nginx.url') || ''
        if (SpmAgent.Config.get('nginx.url') && nginxUrl !== '' && nginxUrl !== 'empty') {
          var secureUrl = nginxUrl.replace(/:.*@/i, ' ')
          console.log('NGINX url: ' + secureUrl)
          var phpFpmUrl = SpmAgent.Config.get('phpFpm.url')
          if (SpmAgent.Config.phpFpm && SpmAgent.Config.phpFpm.url) {
            // this will return an array structure, when defined in config
            phpFpmUrl = SpmAgent.Config.phpFpm.url
          }
          njsAgent.createAgent(new NginxAgent(nginxUrl, null, { phpFpmUrl: phpFpmUrl }))
        } else {
          console.error('Missing nginx status url in config ' + SpmAgent.Config.config)
          process.exit(1)
        }
      }
    } catch (err) {
      console.log(err)
      SpmAgent.Logger.error('Error loading agent ' + err)
    }
  } else {
    nginxUrl = null
    process.exit(1)
  }
  if (autodiscovery) {
    var DockerDiscovery = require('./dockerDiscovery')
    SpmAgent.Logger.info('Watching Docker Events for nginx containers')
    var dockerDiscovery = new DockerDiscovery({
      image: new RegExp(process.env.IMAGE_NAME_PATTERN || 'nginx'),
      startFunction: function (dockerEvent, containerInfo) {
        // console.log(containerInfo.NetworkSettings.IPAddress)
        // console.log(containerInfo.NetworkSettings.Ports['80/tcp'])
        var host = containerInfo.NetworkSettings.IPAddress
        var filterValue = dockerEvent.id.substring(0, 12) // + '/' + containerInfo.NetworkSettings.IPAddress
        var port = 0
        var useSSL = false
        var httpPort = process.env.NGINX_HTTP_PORT || '80/tcp'
        var httpsPort = process.env.NGINX_HTTPS_PORT || '443/tcp'
        var dockerNginxUrl = 'http://' + host + ':' + port + statusPath

        if (process.env.SPM_DOCKER_NETWORK === 'host') {
          SpmAgent.Logger.info('using host network: ' + process.env.SPM_DOCKER_NETWORK)
          if (containerInfo.NetworkSettings.Ports && containerInfo.NetworkSettings.Ports[httpsPort] && containerInfo.NetworkSettings.Ports[httpPort].length > 0) {
            host = containerInfo.NetworkSettings.Ports[httpsPort][0].HostIp
            port = containerInfo.NetworkSettings.Ports[httpsPort][0].HostPort
            useSSL = true
          }
          if (containerInfo.NetworkSettings.Ports && containerInfo.NetworkSettings.Ports[httpPort] && containerInfo.NetworkSettings.Ports[httpPort].length > 0) {
            host = containerInfo.NetworkSettings.Ports[httpPort][0].HostIp
            port = containerInfo.NetworkSettings.Ports[httpPort][0].HostPort
          }
        } else {
          // connect via container network e.g. bridge
          SpmAgent.Logger.info('using container network (' + (process.env.SPM_DOCKER_NETWORK || 'default/bridge') + ')')
          port = (process.env.NGINX_HTTP_PORT || '80/tcp').replace('/tcp', '')
          host = containerInfo.NetworkSettings.IPAddress
        }
        if (useSSL) {
          dockerNginxUrl = 'https://' + host + ':' + port + statusPath
        } else {
          dockerNginxUrl = 'http://' + host + ':' + port + statusPath
        }
        if (port > 0) {
          filterValue = filterValue + '_' + host + ':' + port + statusPath
          SpmAgent.Logger.info('start monitoring: ' + dockerNginxUrl + ' container: ' + filterValue)
          var na = new NginxAgent(dockerNginxUrl, filterValue, { phpFpmUrl: phpFpmStatusPath })
          na = njsAgent.createAgent(na)
          runningNginxAgents[dockerEvent.id] = na
        }
      },
      stopFunction: function (dockerEvent) {
        var na = runningNginxAgents[dockerEvent.id]
        if (na) {
          SpmAgent.Logger.info('stop monitoring for container ' + dockerEvent.id)
          na.stop()
        }
        delete runningNginxAgents[dockerEvent.id]
      }
    })
  }
  return njsAgent
}
NginxMonitor()

process.on('uncaughtException', function (err) {
  console.error((new Date()).toUTCString() + ' uncaughtException:', err.message)
  console.error(err.stack)
  process.exit(1)
})
